### ISSUE TITLE
Maccy window steals focus and pasting fails

### ISSUE THREAD

**Issue by C-Saunders** [p0deje/Maccy#144]
The issue seemed to start when I updated to 0.14.0 earlier today. I'm running MacOS 10.15.6.

I installed 0.13.0 (by downloading the zip file, I usually install via homebrew, but you can't install specific versions of casks) and I am now able to paste from Maccy as expected again.

Thanks for an excellent tool and please let me know if you need any more information I can provide!

**Comment by p0deje** [comment]
Maccy should now steal focus, that's expected. However, it has to return focus when the window is closed so pasting should work just fine.

Can you explain step-by-step what exactly is happening?

**Comment by C-Saunders** [comment]
> Does it fail to paste for all applications or a specific one?

It seems to be happening for all applications. I tried Chrome, VS Code, TextEdit, Slack, and iTerm.

> Can you explain step-by-step what exactly is happening?

- Put my cursor somewhere I can enter text
- Hit hotkeys (cmd + shift + v)
- Maccy window pops up (I have it set to screen center)
- I select an item (mouse, enter key, cmd+#, doesn't seem to change behavior) 
- Focus doesn't return to where it previously was so pasting fails
- The selected item is now the paste-able item and it works if I put focus back where it was and paste normally

Thanks!

**Comment by p0deje** [comment]
Also, any chance you can record a screencast showing the issue? Maybe I can spot something...

**Comment by C-Saunders** [comment]
> Can you try to remove Maccy from "System Preferences" -> "Security & Privacy" -> "Accessibility" and try again? It should ask for permission and work after you give it.

I did this. It asked for permission, I granted it, no change :(

> Can you post the output of `defaults read org.p0deje.Maccy`

```js
{
    "KeyboardShortcuts_popup" = "{\\"carbonModifiers\\":768,\\"carbonKeyCode\\":9}";
    "NSWindow Frame SUUpdateAlert" = "970 768 620 392 0 0 2560 1417 ";
    "NSWindow Frame com.sindresorhus.Preferences.FrameAutosaveName" = "968 502 550 540 0 0 2560 1417 ";
    SUEnableAutomaticChecks = 1;
    SUHasLaunchedBefore = 1;
    SULastCheckTime = "2020-08-07 14:59:04 +0000";
    SUSendProfileInfo = 0;
    SUUpdateRelaunchingMarker = 0;
    fuzzySearch = 1;
    history = {length = 4642931, bytes = 0x62706c69 73743030 af10c800 01000b00 ... 00000000 0046c4c7 };
    historySize = 200;
    hotKey = "command+shift+v";
    ignoredPasteboardTypes =     (
        "com.agilebits.onepassword",
        "de.petermaurer.TransientPasteboardType",
        "com.typeit4me.clipping",
        "Pasteboard generator type"
    );
    maxMenuItems = 28;
    migrations =     {
        "2020-02-22-history-item-add-copied-at" = 1;
        "2020-02-22-history-item-add-number-of-copies" = 1;
        "2020-04-18-switch-storage-to-core-data" = 1;
        "2020-04-25-allow-custom-ignored-types" = 1;
        "2020-06-19-use-keyboardshortcuts" = 1;
    };
    pasteByDefault = true;
    popupPosition = center;
    removeFormattingByDefault = 0;
    showInStatusBar = 1;
}
```

> any chance you can record a screencast showing the issue?

[Screen Recording 2020-08-07 at 1.59.08 PM.mov.zip](https://github.com/p0deje/Maccy/files/5042955/Screen.Recording.2020-08-07.at.1.59.08.PM.mov.zip)

**Comment by p0deje** [comment]
> I did this. It asked for permission, I granted it, no change :(
Just to confirm, have you removed Maccy before trying?

**Comment by C-Saunders** [comment]
> Just to confirm, have you removed Maccy before trying?

I unchecked the box in the Accessibility options, tried to paste via Maccy, it asked for permission, and I granted it.

Do you mean uninstall Maccy, or uncheck the Accessibility box or what I did, or something else?

**Comment by p0deje** [comment]
>I unchecked the box in the Accessibility options, tried to paste via Maccy, it asked for permission, and I granted it.

Not only uncheck, but also remove Maccy by pressing **-** button.

**Comment by C-Saunders** [comment]
Ah, no, I didn't do that. However, I did the whole thing again but including that step just now - no change.

**Comment by p0deje** [comment]
Ok, let me try to reproduce you based on your settings. I'll get back. For now, please revert to 0.13.

**Comment by p0deje** [comment]
I tried setting up the same preferences (see below), also running macOS 10.15.6, but still fail to reproduce :(

1. I'm going to build a special version with additional logging. Will you be able to use it to debug more?
2. Any chance you configured your macOS in some special way? Maybe some accessibility settings are turned?
3. Any chance you run some other app that might interfere somehow? Something like Hammerspoon?

```
{
    "KeyboardShortcuts_popup" = "{\\"carbonModifiers\\":768,\\"carbonKeyCode\\":9}";
    "NSStatusItem Visible Item-0" = 1;
    "NSWindow Frame com.sindresorhus.Preferences.FrameAutosaveName" = "958 368 550 540 0 0 1680 1027 ";
    SUEnableAutomaticChecks = 1;
    SUHasLaunchedBefore = 1;
    SULastCheckTime = "2020-08-07 12:28:16 +0000";
    fuzzySearch = 1;
    history = {length = 42, bytes = 0x62706c69 73743030 a0080000 00000000 ... 00000000 00000009 };
    ignoredPasteboardTypes =     (
        "com.agilebits.onepassword",
        "de.petermaurer.TransientPasteboardType",
        "Pasteboard generator type",
        "com.typeit4me.clipping"
    );
    imageMaxHeight = 16;
    maxMenuItems = 28;
    migrations =     {
        "2020-02-22-history-item-add-copied-at" = 1;
        "2020-02-22-history-item-add-number-of-copies" = 1;
        "2020-04-18-switch-storage-to-core-data" = 1;
        "2020-04-25-allow-custom-ignored-types" = 1;
        "2020-06-19-use-keyboardshortcuts" = 1;
    };
    pasteByDefault = true;
    playSounds = 0;
    popupPosition = center;
    showInStatusBar = 1;
}
```

**Comment by sudo-suhas** [comment]
I faced this issue as well. I had installed a previous version of maccy through brew cask and upgraded from 0.10 to 0.14. Started getting the issue after that. Didn't try to restart the system though.

**Comment by p0deje** [comment]
[Maccy.app.zip](https://github.com/p0deje/Maccy/files/5047812/Maccy.app.zip)

This is a special build with additional logging. Could you try running it and open/paste a few times. Then please open the "Console" app, search for "PROCESS: Maccy" and share the logs.

**Comment by madqueendany** [comment]
hope this helps
```
default	13:03:24.848474+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (before all)
default	13:03:24.848530+0400	Maccy	Return focus: false (before checking visible windows
default	13:03:24.848583+0400	Maccy	Return focus: false (after checking visible windows
default	13:03:24.849476+0400	Maccy	SetFrontProcess: asn=0x0-0x1c9cc9b options=1
default	13:03:25.668782+0400	Maccy	Return focus: false (after closure
default	13:03:25.670193+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (after closure)
default	13:03:26.826632+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (before all)
default	13:03:26.826683+0400	Maccy	Return focus: false (before checking visible windows
default	13:03:26.826734+0400	Maccy	Return focus: false (after checking visible windows
default	13:03:26.827573+0400	Maccy	SetFrontProcess: asn=0x0-0x1c9cc9b options=1
default	13:03:28.771393+0400	Maccy	Accessbility allowed: true
default	13:03:29.010941+0400	Maccy	Return focus: false (after closure
default	13:03:29.011157+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (after closure)
default	13:03:30.472043+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (before all)
default	13:03:30.472099+0400	Maccy	Return focus: false (before checking visible windows
default	13:03:30.472150+0400	Maccy	Return focus: false (after checking visible windows
default	13:03:30.472307+0400	Maccy	SetFrontProcess: asn=0x0-0x1c9cc9b options=1
default	13:03:31.676820+0400	Maccy	Accessbility allowed: true
default	13:03:31.677190+0400	Maccy	Return focus: false (after closure
default	13:03:31.677381+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (after closure)
```

**Comment by p0deje** [comment]
@madqueendany Thanks for the logs! There is special logic in Maccy to **not** return focus to the previous application when any additional Maccy window is opened (e.g. preferences, about). Given you don't have those open, I'd expect the following in logs:

```
default	13:03:26.826683+0400	Maccy	Return focus: false (before checking visible windows
default	13:03:26.826734+0400	Maccy	Return focus: true (after checking visible windows
```

However, your logs clearly show that Maccy believes there is some open window and it should not return focus. I'm pretty sure there is a bug there. I've built one more version which will print all visible windows so hopefully, it will give me a hint of what's wrong. Could you try running it again and sharing logs?

[Maccy.app.zip](https://github.com/p0deje/Maccy/files/5050402/Maccy.app.zip)

**Comment by madqueendany** [comment]
here you are, I'm happy to help you with more logs if you need them, just tell me the steps to set it up

```
default	15:11:29.045722+0400	Maccy	LSExceptions shared instance invalidated for timeout.
default	15:11:40.195997+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (before all)
default	15:11:40.196058+0400	Maccy	Return focus: false (before checking visible windows
default	15:11:40.196141+0400	Maccy	Windows: ["<NSStatusBarWindow: 0x7fc52ae0a730>: \'Item-0\', true", "<NSStatusBarWindow: 0x7fc52ad09d20>: \'\', true"]
default	15:11:40.196186+0400	Maccy	Return focus: false (after checking visible windows
default	15:11:40.196796+0400	Maccy	SetFrontProcess: asn=0x0-0x208a088 options=1
default	15:11:42.396130+0400	Maccy	Accessbility allowed: true
default	15:11:42.396481+0400	Maccy	Return focus: false (after closure
default	15:11:42.396661+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (after closure)
default	15:11:44.642314+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (before all)
default	15:11:44.642364+0400	Maccy	Return focus: false (before checking visible windows
default	15:11:44.642438+0400	Maccy	Windows: ["<NSStatusBarWindow: 0x7fc52ae0a730>: \'Item-0\', true", "<NSStatusBarWindow: 0x7fc52ad09d20>: \'\', true"]
default	15:11:44.642482+0400	Maccy	Return focus: false (after checking visible windows
default	15:11:44.643325+0400	Maccy	SetFrontProcess: asn=0x0-0x208a088 options=1
default	15:11:45.149673+0400	Maccy	Accessbility allowed: true
default	15:11:45.153263+0400	Maccy	Return focus: false (after closure
default	15:11:45.153605+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (after closure)
default	15:11:54.314747+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (before all)
default	15:11:54.314812+0400	Maccy	Return focus: false (before checking visible windows
default	15:11:54.314904+0400	Maccy	Windows: ["<NSStatusBarWindow: 0x7fc52ae0a730>: \'Item-0\', true", "<NSStatusBarWindow: 0x7fc52ad09d20>: \'\', true"]
default	15:11:54.314964+0400	Maccy	Return focus: false (after checking visible windows
default	15:11:54.315954+0400	Maccy	SetFrontProcess: asn=0x0-0x208a088 options=1
default	15:11:55.803890+0400	Maccy	Accessbility allowed: true
default	15:11:55.804234+0400	Maccy	Return focus: false (after closure
default	15:11:55.804429+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (after closure)
default	15:11:57.659838+0400	Maccy	Frontmost app: Optional("com.google.Chrome") (before all)
default	15:11:57.659894+0400	Maccy	Return focus: false (before checking visible windows
default	15:11:57.659970+0400	Maccy	Windows: ["<NSStatusBarWindow: 0x7fc52ae0a730>: \'Item-0\', true", "<NSStatusBarWindow: 0x7fc52ad09d20>: \'\', true"]
default	15:11:57.660017+0400	Maccy	Return focus: false (after checking visible windows
default	15:11:57.660852+0400	Maccy	SetFrontProcess: asn=0x0-0x208a088 options=1
default	15:11:59.273302+0400	Maccy	Accessbility allowed: true
default	15:11:59.273741+0400	Maccy	Return focus: false (after closure
default	15:11:59.273909+0400	Maccy	Frontmost app: Optional("org.p0deje.Maccy") (after closure)
```

**Comment by p0deje** [comment]
Indeed there is one more window that it finds. This is what I see when run this locally:

```
default	07:16:02.261709-0400	Maccy	Windows: ["<NSStatusBarWindow: 0x7fd874008c40>: \'Item-0\', false"]
```

And this is what you have:

```
default	15:11:54.314904+0400	Maccy	Windows: ["<NSStatusBarWindow: 0x7fc52ae0a730>: \'Item-0\', true", "<NSStatusBarWindow: 0x7fc52ad09d20>: \'\', true"]
```

There is 1 more NSStatusBarWindow that is opened so Maccy thinks it should not return the focus back to the previous application. I have zero ideas where this window is coming from but I'll try to make it ignore it. Will share a version to test later today.

**Comment by p0deje** [comment]
[Maccy.app.zip](https://github.com/p0deje/Maccy/files/5050605/Maccy.app.zip)
Ok, I have drafted a patched version. Could you try running it and let me know if it fixes the issue for you? If yes, can you please confirm that when you open Preferences and then open Maccy again, the focus remains on Maccy preferences window?

**Comment by madqueendany** [comment]
The OP's issue is now fixed,

as for your scenario:

1. Made sure that Maccy's active
2. Opened preferences
3. Clicked Google Chrome's address bar
4. Activated Maccy's clipboard window
5. Pasted a random line through a keyboard shortcut
6. Preferences window gets focused.

Hope this helps!

**Comment by p0deje** [comment]
> The OP's issue is now fixed,

Great, I'll polish my quick patch and release 0.14.1 later today.

> as for your scenario:
> 
> 1. Made sure that Maccy's active
> 2. Opened preferences
> 3. Clicked Google Chrome's address bar
> 4. Activated Maccy's clipboard window
> 5. Pasted a random line through a keyboard shortcut
> 6. Preferences window gets focused.
> 
> Hope this helps!

That's an annoying edge case, but I think it's not a big deal at the moment. The other solutions I tried would cause a different edge case when you open preferences, open Maccy again, press Escape, and preferences window is hidden. I haven't found a way to make it work in all edge cases and I believe the current behavior is the least evil.

**Comment by p0deje** [comment]
Please upgrade to https://github.com/p0deje/Maccy/releases/tag/0.14.1, it should be fixed now. If not, please re-open this issue or leave a comment.

Thanks for all the help with debugging!

**Comment by C-Saunders** [comment]
Seems to be working just fine now. Thank you all!

**Comment by tjmcewan** [comment]
Ooh wow, this had me going for half an hour. :) Tried toggling all the settings to no avail; couldn't figure out what I was doing wrong.  Turns out Maccy wasn't listed in A11y settings and it didn't ask to be added.  Great app! :D

**Comment by p0deje** [comment]
Huh, it should have asked for permissions if it didn't.

**Comment by p0deje** [comment]
@FernandoMiguel What's dropdown mode and how Maccy breaks it?

**Comment by FernandoMiguel** [comment]
it's what is called a Hotkey Window
https://www.iterm2.com/documentation-highlights.html

before the window would remain open when accessing maccy history pane. 
now somehow it loses focus and closes, which means once i select something from macy, it is pasting on another app (imagine it ends up on slack)

**Comment by p0deje** [comment]
I guess there is not much I can do about this. Maccy window has to be activated and focused in order for standard language input helpers to work (see #129). Activating Maccy automatically dismisses all context menus on the screen - that's standard AppKit behavior.

**Comment by FernandoMiguel** [comment]
It's strange that this behaviour changed recently.
I believe it was from an update of maccie (tho it could also be an iterm2 update, or even macos beta)
I'll keep looking. But right now, my life is harder, and far less secure

**Comment by p0deje** [comment]
Yes, it did change recently. Maccy started to be force-activated in 0.14.0 via 0e1cb746e7d819eb873d8c0dadcafb3fde347e29

> Support language input helpers when searching (e.g. simplified Chinese input). (#129)

What I'm trying to say is that it's _expected_ that context menus are closed when Maccy popups starting from 0.14.0. I can see how this can be worked around if you create a separate issue with an example of desired behavior.

**Comment by FernandoMiguel** [comment]
Will do thanks clarifying
